src/rgw/driver/d4n/d4n_policy.cc

```diff
@@ -0,0 +1,411 @@
+    return -1;
+  }
+
+  if (!std::get<0>(resp).value()) {
+    if (set_min_avg_weight(0, dir->cct->_conf->rgw_local_cache_address, y)) { /* Initialize minimum average weight */
+      return 0;
+    } else {
+      return -1;
+    }
+  }
+
+  try { 
+    boost::system::error_code ec;
+    response<std::string> value;
+    request req;
+    req.push("HGET", "lfuda", "minAvgWeight:weight");
+      
+    redis_exec(conn, ec, req, value, y);
+
+    if (ec)
+      return -1;
+
+    return std::stoi(std::get<0>(value).value());
+  } catch(std::exception &e) {
+    return -1;
+  }
+}
+
+CacheBlock LFUDAPolicy::find_victim(const DoutPrefixProvider* dpp, rgw::cache::CacheDriver* cacheNode, optional_yield y) {
+  if (entries_map.empty())
+    return {};
+
+  auto it = std::min_element(std::begin(entries_map), std::end(entries_map),
+			      [](const auto& l, const auto& r) { return l.second->localWeight < r.second->localWeight; });
+
+  /* Get victim cache block */
+  std::string key = it->second->key;
+  CacheBlock victim;
+
+  victim.cacheObj.bucketName = key.substr(0, key.find('_')); 
+  key.erase(0, key.find('_') + 1);
+  victim.cacheObj.objName = key.substr(0, key.find('_'));
+  victim.blockID = boost::lexical_cast<uint64_t>(key.substr(key.find('_') + 1, key.length()));
+
+  if (dir->get(&victim, y) < 0) {
+    return {};
+  }
+
+  return victim;
+}
+
+void LFUDAPolicy::shutdown() {
+  dir->shutdown();
+  
+  // call cancel() on the connection's executor
+  boost::asio::dispatch(conn->get_executor(), [c = conn] { c->cancel(); });
+}
+
+int LFUDAPolicy::exist_key(std::string key, optional_yield y) {
+  if (entries_map.count(key) != 0) {
+    return true;
+  }
+
+  return false;
+}
+
+int LFUDAPolicy::get_block(const DoutPrefixProvider* dpp, CacheBlock* block, rgw::cache::CacheDriver* cacheNode, optional_yield y) {
+  response<std::string> resp;
+  int age = get_age(y);
+
+  if (exist_key(block->cacheObj.objName, y)) { /* Local copy */
+    auto it = entries_map.find(dir->build_index(block));
+    it->second->localWeight += age;
+    return cacheNode->set_attr(dpp, block->cacheObj.objName, "localWeight", std::to_string(it->second->localWeight), y);
+  } else {
+    uint64_t freeSpace = cacheNode->get_free_space(dpp);
+
+    while (freeSpace < block->size) { /* Not enough space in local cache */
+      if (int ret = eviction(dpp, cacheNode, y) > 0)
+        freeSpace += ret;
+      else 
+        return -1;
+    }
+
+    int exists = dir->exist_key(block, y);
+    if (exists > 0) { /* Remote copy */
+      if (dir->get(block, y) < 0) {
+	return -1;
+      } else {
+	if (!block->hostsList.empty()) { 
+	  block->globalWeight += age;
+	  
+	  if (dir->update_field(block, "globalWeight", std::to_string(block->globalWeight), y) < 0) {
+	    return -1;
+	  } else {
+	    return 0;
+	  }
+	} else {
+          return -1;
+        }
+      }
+    } else if (!exists) { /* No remote copy */
+      // how to get bufferlist data? -Sam
```

Comment:
For a new local copy, the filter driver is inserting the block in the directory (see handle_data()). Since `insert` will be called after `put_async`, you can update the localWeight of the new block in the `insert` method before adding an entry for it, and then after adding the entry, the attr can be upadted in the cacheNode as well. From the paper `The idea of LFUDA is to adapt to changing workloads by adding
the “age” of the cache to the block value on each access, where the age is defined by
the weight of the last block evicted`, which means the localWeight needs to be updated on every `get` as well. So it makes sense to update the localWeight in `insert`.